### ISSUE TITLE
Recovery command for each check

### ISSUE THREAD

**Issue by runningman84** [sensu/sensu#850]
I would like to have an optional recovery command executed by the sensu client. So I would define check like "fs usage" and a recovery would be "find all old files and delete them". 

Options like recovery timeout and recovery tries would ensure that errors which can not be recovered still create tickets.

**Comment by gwinans** [comment]
See this handler:

https://github.com/nstielau/sensu-community-plugins/blob/remediation/handlers/remediation/sensu.rb

**Comment by runningman84** [comment]
I know the handler but I think the problem could be solved on the client too. Especially for local checks I prefer a local solution.

**Comment by warmfusion** [comment]
write a wrapper command around your check to handle failure cases.

For example:

    {
        "failsafe_check": {
          "command": "my_failsafe.rb check_load.rb args",
          "interval": 60,
          "subscribers": ["..."],
          "handlers": [...],
          "failsafe": {
            "1": {
              "command": "echo 'A Warn happened' && /etc/init.d/nginx restart"
            },
            "2": {
              "command": "echo 'A CRIT happened' && shutdown"
            },
          }
        },
        ...
    }


And then have the script:

1. Run the given arguments as a forked process
2. Capture the exit code and output text
3. Based on the exit code look up the failsafe config and execute the given command for
 the value that matches the exit code

**Comment by portertech** [comment]
I'm not a fan of "auto remediation", as it's really just an indication of technical debt and/or dirty patch work. Disk usage is an interesting example: are old or temporary files no longer important? Can log data be disregarded and deleted, has it been stored elsewhere? Most common "auto remediation" can be better accomplished with things like log rotation or streaming, service supervisors, and bug fixes :+1: 

![1390](https://cloud.githubusercontent.com/assets/149630/7291731/178e0406-e949-11e4-93db-9966e8b9b225.gif)

**Comment by runningman84** [comment]
Ok lets give some examples here:

* Graylog stops working due to "Out of Memory". The process is still up and running but is doesn't process any new message. A log file monitor could catch the error and restart the service.
* A tomcat webapp stops working. A restart of this webapp could help....

Both examples can be coverd my Monit but do we really want to have two different monitoring systems on the same system? But example are appliaction bugs but these bugs can not be fixed by an admin...

**Comment by portertech** [comment]
@runningman84 wouldn't using a proper process supervisor (not monitor) be the best solution for these two examples?

**Comment by runningman84** [comment]
Most applications already come with systemd, upstart or init scripts. We do want to change them. Supervisord doesn't seem to help here. If you give me a process supervisor recommendation I will be happy to check that out. But if a simple monitor like monit contains such feature you might want to add a similar one.

**Comment by warmfusion** [comment]
Perhaps you could use monit alongside sensu;

Have a look at the [monit plugin](https://github.com/sensu/sensu-community-plugins/tree/master/plugins/monit) in the community-plugins project. It appears, at first glance, to provide a mechanism where monit will take care of your services, and when it 'alerts' it'll actually run a sensu check script to report back into the sensu stack.

**Comment by portertech** [comment]
Thinking about possible configuration:

```
"remediation": {
  "command": "true",
  "timeout": 10
}
```

Failure to execute the remediation command (exit 0) could result in a check result with the remediation output.

**Comment by warmfusion** [comment]
it may make more sense to have 'actions' as their own concept; a check could then be configured with 'action handlers' which operate on the client machine and are executed irrespective of what the server wants to do (ie autonomous)

Having the 'action' separate to the check also means 
* it can be easily reused by different checks  (high CPU, low ram, unresponsive tomcat == reboot tomcat)
* it can be recognised as a specific 'action event' by the server (ie Check X failed, and triggered Action Y)
* it helps maintain decoupling of the passive 'observation' from the active 'manipulation' commands in Sensu.

**Comment by runningman84** [comment]
@warmfusion the action concept makes sense to me.
@portertech can you this on your roadmap?

**Comment by jhmartin** [comment]
Any update on this?

**Comment by calebhailey** [comment]
@runningman84 @warmfusion: hi there. Sorry for the radio silence on this. 

I'm going to give this the "Feature" label as part of an ongoing issues triage. No recent internal discussion has happened about this, so this isn't a commitment to actually implement a new feature, but it will flag it for renewed discussion. 

#monitoringlove

**Comment by runningman84** [comment]
@calebhailey any new ideas for a 2.0 design of this feature?

The feature seems to be more important then before if we decide to use standalone checks only in combination with the snssqs transport. A server side solution using handlers would not work here.

**Comment by cjchand** [comment]
+1 for adding an "action" feature, though I fully get - and mostly agree - with @portertech 's point about remediation being a bandaid for crappy code. I've seen remediation cause more issues than it solves, especially when it masks an issue and makes for intermittent behaviors that are very hard to troubleshoot.

One challenge with taking the approach of leaving service restarts to a process management framework: I've seen a few commonly used services don't have proper systemd config as part of their standard packaging (gasp). Example, [Logstash does not ship with any proper systemd config](https://github.com/elastic/logstash/issues/3606)... though they'll be remedying this once 5.0 goes GA, but point still remains. 

This is a problem we've lived with for the past 5+ years: roll your own via supervisord, etc or deal with the shortcomings of the stock process management provided. 

The ugly truth is that processes sometimes get so wrapped around the axle that they need intervention that is more involved than "systemctl restart blah." Given the choice of paging me at 3AM to do it manually vs having Sensu give automated remediation a shot for me first and only waking me up if it fails, I'd take the latter.

Regarding stopgaps, there is the existing ["remediator"](https://github.com/nstielau/sensu-community-plugins/blob/remediation/handlers/remediation/sensu.rb) handler, but that's the community's effort to add what - IMHO - should be a core feature. Likewise, bolting something like Monit on the side of Sensu starts to make things a bit too Rube Goldberg-ish for my liking. I'm all for less moving parts where it makes sense. Even if I wrote something entirely outside of Sensu for remediation, I'd likely end up having it toss events into Sensu anyway. So, why not cut to the chase?
